`Clear-Site-Data: "cache"` HTTP header should clear navigation snapshots

cdumez opened the pull request:
#### 17d23e2ec51d99b28ed99a79de5e0319d23565b7
<pre>
`Clear-Site-Data: &quot;cache&quot;` HTTP header should clear navigation snapshots
<a href="https://bugs.webkit.org/show_bug.cgi?id=251766">https://bugs.webkit.org/show_bug.cgi?id=251766</a>

Reviewed by Geoffrey Garen.

`Clear-Site-Data: &quot;cache&quot;` HTTP header already clears the back/forward cache
but it didn&apos;t clear the navigation snapshots yet. This patch fixes this.

Since a common use case is a log out page, we shouldn&apos;t be able to see
&quot;logged in&quot; information when swiping back once you&apos;ve logged out.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/259936@main">https://commits.webkit.org/259936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48aaf496a48e16f7b68dd4474432b10815242684

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106462 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15485 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115649 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16974 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6710 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98659 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115312 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112228 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95867 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94774 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27511 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8717 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28863 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5435 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14871 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48409 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10798 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3696 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1891,6 +1892,16 @@ void NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin(OptionSet<Web
+#if PLATFORM(COCOA) || PLATFORM(GTK)
+        ViewSnapshotStore::singleton().discardSnapshotImagesForOrigin(origin.topOrigin);
+#endif
+        // Since this navigation requested that we clear existing navigation snapshots, we shouldn't
+        // create a snapshot for this navigation either.
+        if (auto page = WebProcessProxy::webPage(webPageProxyID))
+            page->suppressNextAutomaticNavigationSnapshot();
```
This surprised me a little bit. The rest of the current navigation is still present in the memory cache, right?

cdumez replied:
I don't understand the question.

Let me give an example which hopefully will clarify:
1. Imagine the Page P is getting navigated same-origin from A1 to A2 and A2 serves the `Clear-Site-Data: "cache"` header
2. When we get the resource response for A2, we notice that the Clear-Site-Data is present is we clear all caches for origin A. Note that at this point, we haven't committed the load to A2 yet so the page is still displaying A1. The specification says we have to clear data before we start rendering the response to the user. This means that at this point, we haven't yet created or back/forward cache entry for A1 and we haven't yet snapshotted A1.
3. Once data is cleared, we eventually commit the load once we start getting the response data.

This means that if `Clear-Site-Data: "cache"` didn't suppress the navigation snapshot, a snapshot of A1 would be taken right after A2 asked us to clear caches for origin A. I had to do something similar for the back/forward cache in an earlier patch. It makes sense for snapshot and back/forward cache since they:
1. occur on navigation, after we process the response, when we commit the load
2. apply to the page we navigated AWAY from

Your comment says "the current navigation is still in the memory cache" but navigations are not stored in the memory cache? Also, the memory cache gets cleared before we commit the load of A2.

cdumez replied:
I don't understand the question.

Let me give an example which hopefully will clarify:
1. Imagine the Page P is getting navigated same-origin from A1 to A2 and A2 serves the `Clear-Site-Data: "cache"` header
2. When we get the resource response for A2, we notice that the Clear-Site-Data is present is we clear all caches for origin A. Note that at this point, we haven't committed the load to A2 yet so the page is still displaying A1. The specification says we have to clear data before we start rendering the response to the user. This means that at this point, we haven't yet created or back/forward cache entry for A1 and we haven't yet snapshotted A1.
3. Once data is cleared, we eventually commit the load once we start getting the response data.

This means that if `Clear-Site-Data: "cache"` didn't suppress the navigation snapshot, a snapshot of A1 would be taken right after A2 asked us to clear caches for origin A. I had to do something similar for the back/forward cache in an earlier patch. It makes sense for snapshot and back/forward cache since they:
1. occur on navigation, after we process the response, when we commit the load
2. apply to the page we navigated AWAY from

Your comment says "the current navigation is still in the memory cache" but navigations are not stored in the memory cache? Also, the memory cache gets cleared before we commit the load of A2.

geoffreygaren commented on Source/WebKit/UIProcess/ViewSnapshotStore.cpp:
```diff
@@ -110,6 +111,16 @@ void ViewSnapshotStore::discardSnapshotImages()
+void ViewSnapshotStore::discardSnapshotImagesForOrigin(const WebCore::SecurityOriginData& origin)
+{
+    for (auto it = m_snapshotsWithImages.begin(); it != m_snapshotsWithImages.end();) {
+        auto currentIterator = it;
```
Maybe clearer to say "auto viewSnapshot = *it" here.

cdumez replied:
Ok.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
`Clear-Site-Data: "cache"` HTTP header should clear navigation snapshots
https://bugs.webkit.org/show_bug.cgi?id=251766

Reviewed by Geoffrey Garen.

`Clear-Site-Data: "cache"` HTTP header already clears the back/forward cache
but it didn't clear the navigation snapshots yet. This patch fixes this.

Since a common use case is a log out page, we shouldn't be able to see
"logged in" information when swiping back once you've logged out.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:

Canonical link: https://commits.webkit.org/259936@main
